cookbook/src/tx_segwit-v0.md

```diff
@@ -0,0 +1,351 @@
+#     let sk = SecretKey::new(&mut rand::thread_rng());
+#     let pk = bitcoin::PublicKey::new(sk.public_key(secp));
+#     let wpkh = pk.wpubkey_hash().expect("key is compressed");
+# 
+#     (sk, wpkh)
+# }
+# 
+# 
+# fn receivers_address() -> Address {
+#     Address::from_str("bc1q7cyrfmck2ffu2ud3rn5l5a8yv6f0chkp0zpemf")
+#         .expect("a valid address")
+#         .require_network(Network::Bitcoin)
+#         .expect("valid address for mainnet")
+# }
+# 
+# fn dummy_unspent_transaction_output(wpkh: &WPubkeyHash) -> (OutPoint, TxOut) {
+#     let script_pubkey = ScriptBuf::new_v0_p2wpkh(wpkh);
+# 
+#     let out_point = OutPoint {
+#         txid: Txid::all_zeros(), // Obviously invalid.
+#         vout: 0,
+#     };
+# 
+#     let utxo = TxOut {
+#         value: DUMMY_UTXO_AMOUNT,
+#         script_pubkey,
+#     };
+# 
+#     (out_point, utxo)
+# }
+
+fn main() {
+    let secp = Secp256k1::new();
+    let (sk, wpkh) = senders_keys(&secp);
+    let address = receivers_address();
+    let (dummy_out_point, dummy_utxo) = dummy_unspent_transaction_output(&wpkh);
+
+    // The script code required to spend a p2wpkh output.
+    let script_code = dummy_utxo
+        .script_pubkey
+        .p2wpkh_script_code()
+        .expect("valid script");
+
+    // The input for the transaction we are constructing.
+    let input = TxIn {
+        previous_output: dummy_out_point,
+        script_sig: script_code.clone(),
+        sequence: Sequence::ENABLE_RBF_NO_LOCKTIME,
+        witness: Witness::default(),
+    };
+
+    // The spend output is locked to a key controlled by the receiver.
+    let spend = TxOut {
+        value: SPEND_AMOUNT,
+        script_pubkey: address.script_pubkey(),
+    };
+
+    // The change output is locked to a key controlled by us.
+    let change = TxOut {
+        value: CHANGE_AMOUNT,
+        script_pubkey: ScriptBuf::new_v0_p2wpkh(&wpkh),
+    };
+
+    // The transaction we want to sign and broadcast.
+    let unsigned_tx = Transaction {
+        version: 2,
+        lock_time: absolute::LockTime::from_height(0).expect("valid height"),
+        input: vec![input],
+        output: vec![spend, change],
+    };
+
+    // Sign the unsigned transaction.
+    let mut sighash_cache = SighashCache::new(unsigned_tx);
+    let sighash = sighash_cache
+        .segwit_signature_hash(0, &script_code, DUMMY_UTXO_AMOUNT, EcdsaSighashType::All)
+        .expect("valid sighash");
+    let msg = Message::from(sighash);
+    let sig = secp.sign_ecdsa(&msg, &sk);
+
+    // Update the witness stack
+    let pk = sk.public_key(&secp);
+    sighash_cache
+        .witness_mut(0)
+        .expect("valid witness")
+        .push_bitcoin_signature(&sig.serialize_der(), EcdsaSighashType::All);
+    sighash_cache
+        .witness_mut(0)
+        .expect("valid witness")
+        .push(&pk.serialize());
+
+    // Print the transaction ready to broadcast
+    let tx = sighash_cache.into_transaction();
```

Comment:
While `sighash_cache::witness_mut` this is a fine way to do sighash operations, I feel that it really non-intuitive. It is primarily to avoid allocations when iterating through tx inputs while creating witnesses. 

I feel the below is better for a tutorial/cookbook.

```rust
    // Update the witness wpkh stack: signature + public key.
    let pk = sk.public_key(&secp);
    let mut witness = &mut tx.input[0].witness;

    witness.push_bitcoin_signature(
        &sig.serialize_der(),
        EcdsaSighashType::All
    );
    witness.push(&pk.serialize());
```

Reply:
I've updated how we are mutating the witness in both segwit and taproot with your suggestion.
Can you check it now please?